File: Leetcode/Daily/07-2025/Python3/06.py
# Problem: 1865. Finding Pairs With a Certain Sum (Medium)
# Link: https://leetcode.com/problems/finding-pairs-with-a-certain-sum

class FindSumPairs:

    def __init__(self, nums1: List[int], nums2: List[int]):
        self.v = []
        self.umap1 = {}
        self.umap2 = {}

        for i in nums1:
            self.umap1[i] = self.umap1.get(i, 0) + 1
        
        for i in nums2:
            self.umap2[i] = self.umap2.get(i, 0) + 1

        self.v = nums2
                

    def add(self, index: int, val: int) -> None:
        self.umap2[self.v[index]] -= 1
        self.v[index] += val
        self.umap2[self.v[index]] = self.umap2.get(self.v[index], 0) + 1
                

    def count(self, tot: int) -> int:
        ans = 0

        for key in self.umap1:
            if tot - key in self.umap2:
                ans += self.umap1[key] * self.umap2[tot - key]

        return ans
        


# Your FindSumPairs object will be instantiated and called as such:
# obj = FindSumPairs(nums1, nums2)
# obj.add(index,val)
# param_2 = obj.count(tot)

File: Leetcode/Daily/07-2025/TypeScript/06.ts
// Problem: 1865. Finding Pairs With a Certain Sum (Medium)
// Link: https://leetcode.com/problems/finding-pairs-with-a-certain-sum

class FindSumPairs {
    v = [];
    umap1 = new Map<number, number>();
    umap2 = new Map<number, number>();

    constructor(nums1: number[], nums2: number[]) {
        for (let i of nums1) this.umap1.set(i, (this.umap1.get(i) || 0) + 1);
        for (let i of nums2) this.umap2.set(i, (this.umap2.get(i) || 0) + 1);

        this.v = nums2;
    }

    add(index: number, val: number): void {
        this.umap2.set(this.v[index], this.umap2.get(this.v[index]) - 1);
        this.v[index] += val;
        this.umap2.set(this.v[index], (this.umap2.get(this.v[index]) || 0) + 1);
    }

    count(tot: number): number {
        let ans = 0;

        for (let key of this.umap1.keys()) {
            if (this.umap2.has(tot - key)) {
                ans += this.umap1.get(key) * this.umap2.get(tot - key);
            }
        }

        return ans;
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * var obj = new FindSumPairs(nums1, nums2)
 * obj.add(index,val)
 * var param_2 = obj.count(tot)
 */